### PR TITLE
feat(go): support go.mod as idiomatic version file

### PR DESCRIPTION
**Pull request by JamBalaya56562**
## Summary

Reads the Go version from `go.mod` as an idiomatic version file, addressing
discussion [#4136](https://github.com/jdx/mise/discussions/4136).

This was first added in #4312 and **reverted 4 days later** (`842d051` → `7fc9beb`). This PR re-implements it in a way that resolves both root causes of that revert.

## Why it was reverted before (and why this is different)

1. **The `go` directive is a *minimum* version, not a pin.** #4312 used `go 1.14` verbatim as the version to select, so mise forced users onto the module's floor version instead of their intended/latest one ([confirmed by a user on the revert day](https://github.com/jdx/mise/discussions/4136)).
2. **Idiomatic version files were enabled by default then**, so the change silently affected *every* Go user with a `go.mod` — a widespread regression, hence the fast revert.

Both are addressed:

- **Opt-in.** Idiomatic version files are disabled by default (#6501) and enabled per-tool (#4902). `go.mod` is only read when the user runs `mise settings add idiomatic_version_file_enable_tools go`, so unlike #4312 it cannot silently change behavior for everyone.
- **Correct version semantics.** The `toolchain goX.Y.Z` directive (an exact pin — what `go version` reports in the repo) takes precedence when present and resolves exactly. Otherwise the `go X.Y` minimum resolves as a **prefix to the latest matching patch** (e.g. `go 1.22` → latest `1.22.x`), consistent with how every other idiomatic version file is resolved (no self-sorting; the backend resolves the request string). Pre-releases and malformed values fail safe (the file is skipped).

## Changes

- `src/plugins/core/go.rs`: add `go.mod` to `_idiomatic_filenames`, override `_parse_idiomatic_file`, add `parse_gomod`/`is_go_version`, and a `#[cfg(test)]` unit test covering `toolchain` precedence, `toolchain default` fallback, full-patch `go`, comments/whitespace, pre-releases, and malformed input. Modeled on the Ruby `Gemfile` plugin as suggested in the discussion.
- `e2e/core/test_go_from_gomod`: opt-in e2e test (toolchain precedence + minimum→latest-patch). The "ignored without opt-in" case remains covered by `test_go_slow`.
- `docs/configuration.md`: add `go.mod` to the idiomatic files table and document the minimum-version semantics.

## Notes

- Behavior is unchanged unless a user explicitly opts in for `go`.
- Full build / clippy / e2e validated in CI.

*This PR was prepared by an AI coding assistant.*

## Summary by CodeRabbit

* **New Features**
  * Go projects can now use `go.mod` as an idiomatic version source alongside existing idiomatic version files.
  * `toolchain goX.Y.Z` takes priority when determining the Go version.
  * If only `go X.Y` is present (or `toolchain default` is used), it resolves to the latest matching patch release.
  * Invalid or unsupported directives are ignored.
* **Documentation**
  * Updated configuration guidance to explain `go.mod` directive interpretation and pin vs minimum behavior.
* **Tests**
  * Added end-to-end and unit test coverage for precedence, fallback behavior, and comment/whitespace and invalid-format handling.

### CODE REVIEW

File: src/plugins/core/go.rs [hunk 1]
```diff
@@ -394,3 +409,96 @@ impl Backend for GoPlugin {
         })
     }
 }
+
+/// Returns true if `v` is a plain numeric-dotted version like `1.22` or `1.22.5`.
+/// Pre-releases (`1.22rc1`), `default`, and other non-numeric forms are rejected so
+/// they never mis-resolve.
+fn is_go_version(v: &str) -> bool {
+    regex!(r"^[0-9]+(\.[0-9]+)*$").is_match(v)
+}
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/plugins/core/go.rs]: Good catch — fixed in `e073257`. Split the validator so the `go` directive now requires `major.minor` with an optional patch (`^[0-9]+\.[0-9]+(\.[0-9]+)?$`) via `is_go_directive_version`, so a malformed `go 1` no longer parses to a `1` prefix that could resolve to the newest Go release. Added a `parse_gomod("go 1\n") == ""` test case.

*This comment was generated by an AI coding assistant.*

File: src/plugins/core/go.rs [hunk 2]
```diff
@@ -394,3 +409,96 @@ impl Backend for GoPlugin {
...
+    };
+
+    // Prefer a concrete `toolchain goX.Y.Z` pin; otherwise fall back to the `go` minimum.
+    // An invalid toolchain value (e.g. `toolchain default`, a pre-release) falls through
+    // to a usable `go` directive rather than discarding the file.
+    directive_value("toolchain")
+        .and_then(|v| v.strip_prefix("go").map(|s| s.to_string()))
+        .filter(|v| is_go_version(v))
+        .or_else(|| directive_value("go").filter(|v| is_go_version(v)))
+        .unwrap_or_default()
```
[the remark this answers is not in the document]
Comment (JamBalaya56562) [reply src/plugins/core/go.rs]: Addressed in `e073257`. The `toolchain` value must now be a fully-qualified `goX.Y.Z` release (`is_go_toolchain_version`) to be used as an exact pin; partial (`toolchain go1.22`), pre-release, or `default` values fall back to the `go` directive. Added tests for the partial toolchain, toolchain-only, and bare `go 1` cases.

I intentionally did **not** add a numeric `toolchain >= go` comparison: a toolchain lower than the `go` directive is a go.mod that the `go` command itself rejects, so it can't occur in valid files, and enforcing it would require an inline Go-version ordering comparison — something this repo's guidelines steer away from (`CLAUDE.md`: "DO NOT ASSUME SEMVER" — avoid new version-comparison call sites for picking versions). Happy to revisit if you'd prefer it enforced.

(Separately, the e2e `assert_contains` nitpick from this review is also fixed — the test now derives the expected patch via `mise ls-remote go@1.21 | tail -1` and asserts exact equality.)

*This comment was generated by an AI coding assistant.*
